### PR TITLE
Add coercion support to fields

### PR DESCRIPTION
**Pull request by zisten**
The code is based on [Type::Coercion Inlining Methods](https://metacpan.org/pod/Type::Coercion#Inlining-methods). Coercion feature for fields added with a simple test. 

See more `inlined` details in [Type::Tiny `inlined` attribute](https://metacpan.org/pod/Type::Tiny#inlined)

### CODE REVIEW

**Comment by zisten** [comment]
I don't hnow why it still failed, but this will work for me. May someone fix the failing?

File: lib/Mandel/Model.pm [hunk 1]
```diff
@@ -153,8 +153,11 @@ sub _field_type {
 
   use Types::Standard qw( Num );
 
+  if ($type->has_coercion and $type->coercion->can_be_inlined) {
+    $code .= '$_ = '.$type->coercion->inline_coercion('$_').";\n";
+  }
   if ($type->can_be_inlined) {
-    $code .= $type->inline_assert('$_');
+    $code .= $type->inline_assert('$_')."\n";
```
Comment (jhthorsen) [inline lib/Mandel/Model.pm]: Why do you need newlines? I would believe just ```;``` should be enough.

File: .travis.yml [hunk 2]
```diff
@@ -7,9 +7,9 @@ env:
   - "HARNESS_OPTIONS=j4 TEST_ONLINE=1"
 before_install:
   - sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10
-  - echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' | sudo tee /etc/apt/sources.list.d/mongodb.list
+  - echo 'deb http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.0 multiverse' | sudo tee /etc/apt/sources.list.d/mongodb-org-3.0.list
   - sudo apt-get update -q
-  - sudo apt-get install mongodb-org-server
+  - sudo apt-get install -y mongodb-org-server
```
Comment (jhthorsen) [inline .travis.yml]: This change does not belong in this pull request.

**Comment by jhthorsen** [comment]
It's difficult to figure out, since https://github.com/jhthorsen/mandel/blob/master/t/collection-online.t#L16 is messed up. It should have been something like this:
```
  my ($col, $err, $doc);
  $collection->save(
    {name => 'Bruce'},
    sub { ($col, $err, $doc) = @_; $id = $doc->id; Mojo::IOLoop->stop; }
  );
  Mojo::IOLoop->start;

  is $col, $collection, 'got collection';
  ok !$err, 'next: no error' or diag "save bruce failed: $err";
  isa_ok $doc, 'Mandel::Document';
  is $doc->name, 'Bruce', 'saved bruce';
```

**Comment by jhthorsen** [comment]
And it's probably not your fault, since https://travis-ci.org/jhthorsen/mandel/jobs/52638754 is failing as well (I just re-ran it). When that is said: I can't merge with failing tests.

File: lib/Mandel/Model.pm (hunk 1 above)
Comment (zisten) [inline lib/Mandel/Model.pm]: Yes, it's enought, but I add newlines like other codes do:

```perl
    $code .= "local \$_ = \$_[1];\n";
    $code .= $self->_field_type($meta->{isa}) if $meta->{isa};
    $code .= "\$_[0]->{dirty}{$name} = 1;";
    $code .= "\$raw->{'$name'} = \$_;\n";
    $code .= "return \$_[0];\n}";
```

Comment (jhthorsen) [inline lib/Mandel/Model.pm]: Ok. That's good enough for me :+1:

**Comment by zisten** [comment]
@jhthorsen, the failing is caused by Mango - because the version 1.7 requires Mongodb 3.0 for work, force to install Mango 1.6 should fix it - but it's damn.

I will push another commit for this, just for minute.

**Comment by jhthorsen** [comment]
Shouldn't that be fixed by b825c887e715a23e78766d8ecffda8b7e66053df?

**Comment by zisten** [comment]
I also think so, but failed - that's strange, maybe the builtin mongodb service in travis.ci caused this failing?

**Comment by zisten** [comment]
I have post a fix commit for the faling build, the build is passed sure, but it should have a better method for it.

**Comment by jhthorsen** [comment]
I don't think so. https://travis-ci.org/jhthorsen/mandel/jobs/58377728 says:
```
$ sudo apt-get install mongodb-org-server
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libgeos-3.2.2
Use 'apt-get autoremove' to remove them.
The following packages will be REMOVED:
  mongodb-10gen
The following NEW packages will be installed:
  mongodb-org-server
0 upgraded, 1 newly installed, 1 to remove and 191 not upgraded.
Need to get 9,521 kB of archives.
After this operation, 201 MB disk space will be freed.
Get:1 http://downloads-distro.mongodb.org/repo/ubuntu-upstart/ dist/10gen mongodb-org-server amd64 2.6.9 [9,521 kB]
Fetched 9,521 kB in 0s (17.1 MB/s)
(Reading database ... 75363 files and directories currently installed.)
Removing mongodb-10gen ...
arg: remove
 * Stopping database mongodb       Processing triggers for ureadahead ...
Processing triggers for man-db ...
Selecting previously unselected package mongodb-org-server.
(Reading database ... 75332 files and directories currently installed.)
Unpacking mongodb-org-server (from .../mongodb-org-server_2.6.9_amd64.deb) ...
Processing triggers for man-db ...
Processing triggers for ureadahead ...
Setting up mongodb-org-server (2.6.9) ...
mongod start/running, process 3422
```

Hm... When that is said: It also says that it's installing ```mongodb-org-server_2.6.9```

**Comment by jhthorsen** [comment]
Could you try to figure out why it installs 2.6.9, before I hit the "Merge" button..?

**Comment by zisten** [comment]
In this repo the latest version is 2.6.9
```yaml
before_install:
  - sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10
  - echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' | sudo tee /etc/apt/sources.list.d/mongodb.list
  - sudo apt-get update -q
  - sudo apt-get install mongodb-org-server
```

Mango 1.7 requires the mongodb version 3.0, which the repo is:
```
 echo "deb http://repo.mongodb.org/apt/ubuntu "$(lsb_release -sc)"/mongodb-org/3.0 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-3.0.list
```

I've tried mongodb 3.0 but failed. Maybe we should wait for a newer version of Mango for a better fix.

**Comment by jhthorsen** [comment]
Ok. Thanks for trying. Merging and releasing now.

**Comment by jhthorsen** [comment]
Thanks for the contribution.

https://metacpan.org/release/JHTHORSEN/Mandel-0.26

**Comment by zisten** [comment]
:+1:
